src/components/camera/CameraStream.js

```diff
@@ -26,15 +26,83 @@ function CameraStream({ cameraName }) {
 
   const [lastFrameTime, setLastFrameTime] = useState(0.0);
   const [currentFpsAvg, setCurrentFpsAvg] = useState(20);
-  const vidTag = <video id={`${cameraName}-player`} className='video-tag' muted autoPlay preload="auto" alt={`${cameraTitle} stream`}></video>;
+  const [popoutWindow, setPopoutWindow] = useState(null);
+  const [aspectRatio, setAspectRatio] = useState(1);
+
+  let cameraCanvas = useRef(null);
+  let cameraContext = useRef(null);
+  
+  const vidTag = useMemo(() => {
+    return <video style={{opacity: popoutWindow ? '0' : '1'}} id={`${cameraName}-player`} className='video-tag' muted autoPlay preload="auto" alt={`${cameraTitle} stream`}></video>;
+  }, [cameraName, cameraTitle, popoutWindow])
+
+  const handlePopOut = useCallback(() => {
+    if (popoutWindow) {
+      // if the window popout exists
+      popoutWindow.close();
+      setPopoutWindow(null);
+    } else {
+      // if the window popout doesn't exist
+      let newWindow = window.open("", "", "width=500,height=500");
+      newWindow.document.body.style.margin = "0";
+      newWindow.document.title = cameraTitle + " Stream";
+      newWindow.document.body.innerHTML = `
+      <div style="background-color:black;width:100%;height:100%;font-family:Arial,Helvetica,sans-serif;display:flex;justify-content:center">
+        <div style="z-index:100;position:absolute;width:100%;font-size:30px;font-weight:bold;text-align:center;color:white;padding-top:5px;padding-bottom:5px;background-color:#00000066;">${cameraTitle}</div>
+        <div id="ext-fps" style="z-index:101;position:absolute;top:5;left:5;font-size:15px;font-weight:bold;color:red;">FPS: N/A</div>
+        <canvas id="ext-vid" style="z-index:1;border:none;display:block;margin:auto 0;"></canvas>
+        </div>
+      </div>`;
+
+      let canvas = newWindow.document.querySelector('#ext-vid');
+      let context = canvas.getContext('2d');
+      let aspectRatio = document.querySelector(`#${cameraName}-player`).videoHeight / document.querySelector(`#${cameraName}-player`).videoWidth;
+      setAspectRatio(aspectRatio);
+      
+      canvas.width = aspectRatio * 400;
+      canvas.height = 400;
+      context.fillStyle = "black";
+      context.fillRect(0, 0, canvas.width, canvas.height);
+      
+      setPopoutWindow(newWindow);
+      window.onunload = function() {
+        if (newWindow && !newWindow.closed) {
+            newWindow.close();
+        }
+      };
+      cameraCanvas.current = canvas;
+      cameraContext.current = context;
+      newWindow.addEventListener("beforeunload", () => {
+        setPopoutWindow(null);
+      });
```

Comment:
This is pretty long and makes the component hard to read overall. Can you factor this out into a method defined outside the component?